src/Workspaces/Remote/Core/Diagnostics/PerformanceTrackerService.cs

```diff
@@ -0,0 +1,445 @@
+
+                var analyzerCount = kNeighborIndices.Count;
+                for (var index = 0; index < analyzerCount; index++)
+                {
+                    var analyzerId = analyzerIdIndex[index];
+
+                    // if result performance is lower than our threshold, don't need to calcuate
+                    // LOF value for the analyzer
+                    var rawData = rawPerformanceData[analyzerId];
+                    if (rawData.average <= _averageThreshold && rawData.stddev <= _stddevThreshold)
+                    {
+                        continue;
+                    }
+
+                    // possible bad analyzer, calcuate LOF
+                    var lof_value = TryGetLocalOutlierFactor(allDistances, kNeighborIndices, kDistances, index);
+                    if (!lof_value.HasValue)
+                    {
+                        // this analyzer doesn't have lof value
+                        continue;
+                    }
+
+                    if (lof_value <= _minLOFValue)
+                    {
+                        // this doesn't stand out from other analyzers
+                        continue;
+                    }
+
+                    // report found possible bad analyzers
+                    _badAnalyzers.Add(new ExpensiveAnalyzerInfo(_owner.AllowTelemetry(analyzerId), analyzerId, lof_value.Value, rawData.average, rawData.stddev));
+                }
+
+                _badAnalyzers.Sort(this);
+            }
+
+            private double? TryGetLocalOutlierFactor(
+                List<List<double>> allDistances, List<List<int>> kNeighborIndices, List<double> kDistances, int analyzerIndex)
+            {
+                var rowKNeighborsIndices = kNeighborIndices[analyzerIndex];
+                if (rowKNeighborsIndices.Count == 0)
+                {
+                    // nothing to calculate if there is no neighbor to compare
+                    return null;
+                }
+
+                var lrda = TryGetLocalReachabilityDensity(allDistances, kNeighborIndices, kDistances, analyzerIndex);
+                if (!lrda.HasValue)
+                {
+                    // can't calculate reachability for the analyzer. can't calculate lof for this analyzer
+                    return null;
+                }
+
+                var lrdb = 0D;
+                foreach (var neighborIndex in rowKNeighborsIndices)
+                {
+                    var reachability = TryGetLocalReachabilityDensity(allDistances, kNeighborIndices, kDistances, neighborIndex);
+                    if (!reachability.HasValue)
+                    {
+                        // this neighbor analyzer doesn't have its own neighbor. skip it
+                        continue;
+                    }
+
+                    lrdb += reachability.Value;
+                }
+
+                return lrdb / rowKNeighborsIndices.Count / lrda;
```

Comment:
:bulb: Please add parentheses around the appropriate sub-expression.

Reply:
not sure what you mean? it is all divide. doesn't matter where parentheses are.